storage/innobase/fil/fil0fil.cc

```diff
@@ -5927,9 +5923,14 @@ dberr_t fil_ibd_open(bool validate, fil_type_t purpose, space_id_t space_id,
   ut_ad(!is_encrypted || df.is_open());
 
   const byte *first_page = df.is_open() ? df.get_first_page() : nullptr;
-  fil_space_crypt_t *crypt_data =
-      first_page ? fil_space_read_crypt_data(page_size_t(flags), first_page)
-                 : nullptr;
+  fil_space_crypt_t *crypt_data{nullptr};
+  if (df.m_crypt_data) {
+    // crypt data was already read by validate_first_page
+    crypt_data = df.m_crypt_data;
+    df.m_crypt_data = nullptr;
```

Comment:
How does this work? This is a can of worms.

df is destroyed after the initial startup phase. crypt_data is destroyed along with it.  So fil_space_t will point to dangling pointer (df.m_crypt_data)

Another issue is when freeing this dangling pointer.

Reply:
Can we avoid reading crypt_data at this stage from df? Remember, we are reading from non-redo applied page. Better to read page 0 from buffer pool and get crypt-data.

Reply:
There was no dangling pointers - please note that after assigning m_crypt_data to space - I have made df.m_crypt_data = nullptr. So df destructor would not remove the, now space's, crypt_data. However, you are right about non-redo applied page. I have missed that. I have removed the m_crypt_data from df and went back with reading the crypt_data from page0.

Reply:
Datafile objects lifetime is very less and to transfer data to the proper life time objects should be one in proper way. THis is risky and error prone. We avoided the pitfall for now..

Thanks for moving to page 0 reading.